### ISSUE TITLE
Unable to read/write unicode string arrays

### ISSUE THREAD

**Issue by c-mita** [h5py/h5py#624]
A TypeError is raised when trying to create a dataset using an array of unicode strings.

Running the following:

```
import h5py

f = h5py.File("test.h5", 'w')
a = f.create_group("a")
a.create_dataset("b", data = [u"testdata"])
```

Results in

```
Traceback (most recent call last):
  File "/home/mew09119/h5pytest.py", line 5, in <module>
    a.create_dataset("b", data=[u"testdata"])
  File "/scratch/h5py/h5py/_hl/group.py", line 105, in create_dataset
    dsid = dataset.make_new_dset(self, shape, dtype, data, **kwds)
  File "/scratch/h5py/h5py/_hl/dataset.py", line 92, in make_new_dset
    tid = h5t.py_create(dtype, logical=1)
  File "h5py/h5t.pyx", line 1448, in h5py.h5t.py_create (/scratch/h5py/h5py/h5t.c:15481)
  File "h5py/h5t.pyx", line 1468, in h5py.h5t.py_create (/scratch/h5py/h5py/h5t.c:15312)
  File "h5py/h5t.pyx", line 1529, in h5py.h5t.py_create (/scratch/h5py/h5py/h5t.c:15272)
TypeError: No conversion path for dtype: dtype('<U8')
```

This works fine if the `u` is removed from before the `"testdata"`

Trying to read a dataset of fixed length unicode strings also fails.
For example, given the following file (shown with H5Dump):

```
HDF5 "test.h5" {
GROUP "/" {
   GROUP "a" {
      GROUP "b" {
         ATTRIBUTE "test" {
            DATATYPE  H5T_STRING {
               STRSIZE 22;
               STRPAD H5T_STR_NULLTERM;
               CSET H5T_CSET_UTF8;
               CTYPE H5T_C_S1;
            }
            DATASPACE  SIMPLE { ( 2, 2 ) / ( 2, 2 ) }
            DATA {
            (0,0): "A String", "String1",
            (1,0): "Some\37777777742\37777777604\37777777653ttributeString",
            (1,1): "String2"
            }
         }
      }
   }
}
}
```

Trying to read the attribute `"test"` (using `f["a"]["b"].attrs["test"]`) results in:

```
Traceback (most recent call last):
  File "/home/mew09119/h5pytest.py", line 7, in <module>
    f["a"]["b"].attrs["test"]
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper (/scratch/h5py/h5py/_objects.c:2456)
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper (/scratch/h5py/h5py/_objects.c:2413)
  File "/home/mew09119/.local/lib/python2.7/site-packages/h5py/_hl/attrs.py", line 79, in __getitem__
    attr.read(arr, mtype=htype)
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper (/scratch/h5py/h5py/_objects.c:2456)
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper (/scratch/h5py/h5py/_objects.c:2413)
  File "h5py/h5a.pyx", line 355, in h5py.h5a.AttrID.read (/scratch/h5py/h5py/h5a.c:4999)
  File "h5py/_proxy.pyx", line 36, in h5py._proxy.attr_rw (/scratch/h5py/h5py/_proxy.c:899)
IOError: Unable to read attribute (No appropriate function for conversion path)
```

**Comment by JohnHBrock** [comment]
This is still an issue as of v2.5.0.

**Comment by aragilar** [comment]
This is effectively the same as https://github.com/h5py/h5py/issues/289, with the same workaround.

**Comment by t-b** [comment]
Can be closed I guess as #289 is closed as well.

**Comment by aragilar** [comment]
Closed by #1032.

**Comment by ngeraedts** [comment]
This particular issue still exists in v2.10.0. #1032 fixes the issue when saving a list of unicode values on attributes, but the problem persists when calling `create_dataset('f', data=[u'hello'])`.
